### ISSUE TITLE
PostgreSQL segfaults

### ISSUE THREAD

**Issue by strk** [larsop/resolve-overlap-and-gap#1]
@larsop reported backend crashes during PostGIS Topology operations.
This ticket is to track the work of providing smallest testcase that can trigger such behavior
and fix PostGIS (or PostgreSQL) to stop doing that.

It will ALWAYS be a bug in whatever core or C extension of PostgreSQL to crash a backend, as in no occurrence should procedural code affect a PostgreSQL cluster so badly.

Problem was reported to occur with both PostgreSQL 12 and 11 with PostGIS 2.5.3 and GEOS 3.8.0.

The current scenario triggering the crash involves multiple threads (I think that means multiple backends) trying heal borders that are connected or close using topology.ST_ModEdgeHeal.

**Comment by strk** [comment]
It's to be noted that the latest release in the 2.5 series of PostGIS is 2.5.4 (crash in report is with 2.5.3).
In the [CHANGELOG](https://git.osgeo.org/gitea/postgis/postgis/src/tag/2.5.4/NEWS) for that version there's a [mention](https://git.osgeo.org/gitea/postgis/postgis/src/tag/2.5.4/NEWS#L16) of a crash fix which may be a possible cause for a random crash: https://trac.osgeo.org/postgis/ticket/4519 -- it is reported to occur when query + srs > 256 chars

**Comment by larsop** [comment]
I failed to install postgis 2.5.4. I tried to remove all Postgres/Postgis and installed this cmd 

`
yum install postgresql12-debug postgresql12-server postgresql12-contrib  postgresql12-docs postgis25_12 postgis25_12-utils postgis25_12-client postgis25_12-docs pgrouting_12
`
and git this error
`
file /usr/pgsql-12/lib/postgis-2.5.so conflicts between attempted installs of postgis30_12-3.0.1-4.rhel7.x86_64 and postgis25_12-2.5.4-1.rhel7.x86_64
  file /usr/pgsql-12/lib/postgis_topology-2.5.so conflicts between attempted installs of postgis30_12-3.0.1-4.rhel7.x86_64 and postgis25_12-2.5.4-1.rhel7.x86_64
  file /usr/pgsql-12/share/extension/address_standardizer.control conflicts between attempted installs of postgis30_12-3.0.1-4.rhel7.x86_64 and postgis25_12-2.5.4-1.rhel7.x86_64
  file /usr/pgsql-12/share/extension/address_standardizer.sql conflicts between attempted installs of postgis30_12-3.0.1-4.rhel7.x86_64 and postgis25_12-2.5.4-1.rhel7.x86_64
  file /usr/pgsql-12/share/extension/address_standardizer_data_us.control conflicts between attempted installs of postgis30_12-3.0.1-4.rhel7.x86_64 and postgis25_12-2.5.4-1.rhel7.x86_64
  file /usr/pgsql-12/share/extension/postgis.control conflicts between attempted installs of postgis30_12-3.0.1-4.rhel7.x86_64 and postgis25_12-2.5.4-1.rhel7.x86_64
  file /usr/pgsql-12/share/extension/postgis_sfcgal.control conflicts between attempted installs of postgis30_12-3.0.1-4.rhel7.x86_64 and postgis25_12-2.5.4-1.rhel7.x86_64
  file /usr/pgsql-12/share/extension/postgis_tiger_geocoder.control conflicts between attempted installs of postgis30_12-3.0.1-4.rhel7.x86_64 and postgis25_12-2.5.4-1.rhel7.x86_64
  file /usr/pgsql-12/share/extension/postgis_topology.control conflicts between attempted installs of postgis30_12-3.0.1-4.rhel7.x86_64 and postgis25_12-2.5.4-1.rhel7.x86_64
`
So I installed Postgis 3.0  instead

**Comment by larsop** [comment]
Running with 
` POSTGIS="3.0.1 ec2a9aa" [EXTENSION] PGSQL="120" GEOS="3.8.1-CAPI-1.13.3" PROJ="7.0.0" LIBXML="2.9.1" LIBJSON="0.11" LIBPROTOBUF="1.0.2" WAGYU="0.4.3 (Internal)" TOPOLOGY
`
it still core dumped. I was running 20 threads
`"CALL resolve_overlap_gap_run(
('sl_kbj.trl_2019_test_segmenter_mindredata','gid','geo',25833,true),
('topo_sr16_mdata_01',1.0),
resolve_overlap_data_clean_type_func(49,5,10000,3,25,120,240,35,325),
20,2500)"`

In the log i see this

`NOTICE:  cell 199 cell_job_type 2, has_edges t, _loop_number 1
< , 2020-04-21 04:52:02 CEST, , 2020-04-21 04:52:02.366 CEST >LOG:  server process (PID 23304) was terminated by signal 11: Segmentation fault
< , 2020-04-21 04:52:02 CEST, , 2020-04-21 04:52:02.366 CEST >DETAIL:  Failed process was running: CALL resolve_overlap_gap_single_cell(
          'sl_kbj.trl_2019_test_segmenter_mindredata','geo','gid','topo_sr16_mdata_01.trl_2019_test_segmenter_mindredata',
          'topo_sr16_mdata_01',1,25833,'true',
          '(49,5,10000,3,25,120,240,3,35)',
          'topo_sr16_mdata_01.trl_2019_test_segmenter_mindredata_job_list','topo_sr16_mdata_01.trl_2019_test_segmenter_mindredata_grid','0103000020E964000001000000050000000000000018230F41E5354051B6C95A410000000018230F41C452C98E9BCE5A41000000002CD70F41C452C98E9BCE5A41000000002CD70F41E5354051B6C95A410000000018230F41E5354051B6C95A41',2,1);
< , 2020-04-21 04:52:02 CEST, , 2020-04-21 04:52:02.366 CEST >LOG:  terminating any other active server processes
< postgres, 2020-04-21 04:52:02 CEST, resolve_cha, 2020-04-21 04:52:02.366 CEST >WARNING:  terminating connection because of crash of another server process`
and
`Apr 21 04:52:02 vroom2 kernel: postmaster[23304]: segfault at fffffffffffffff8 ip 00000000008a14d3 sp 00007fffb00577a8 error 5 in postgres[400000+72f000]
Apr 21 04:52:02 vroom2 kernel: postmaster[23294]: segfault at fffffffffffffff8 ip 00000000008a14d3 sp 00007fffb00577a8 error 5 in postgres[400000+72f000]
`

**Comment by larsop** [comment]
The merged the code that fails into  master now.

**Comment by strk** [comment]
Which commit / where do I find instructions to try at reproducing it ?
Can you reproduce with single thread ?

**Comment by larsop** [comment]
It has been problem a long time I can not identify a certain commit.

I have never seen it with a single thread. If I let work at cell with common lines from simple feature work in parallel at the same time it seems happen more often.

**Comment by strk** [comment]
> It has been problem a long time I can not identify a certain commit.

I was referring to git commit, in replay to https://github.com/larsop/resolve-overlap-and-gap/issues/1#issuecomment-616977427  (sorry for the confusion) :)

> I have never seen it with a single thread.

Ok, this is good info. So we're looking for a problem related to multi-threading, or I guess it's really multi-process because dblink will create another PostgreSQL backend process for each of what you call "threads", is this correct ?

> If I let work at cell with common lines from simple feature work in parallel at the same time it seems happen more often.

Ok, for the sake of having a testcase then we WANT to do this, we need something that is much more likely to segfault. Can you provide such testcase ? Ideally it would also work with PostgreSQL 9.6, to test multiple PostgreSQL versions with it.

**Comment by larsop** [comment]
Ok, I will try yo make a test case that fails with small dataset and  not running 20 threads in parallel.

**Comment by larsop** [comment]
I now made branch https://github.com/larsop/resolve-overlap-and-gap/tree/postgres_core_dump
 and a small data set that core dumped 4 off 5  times after 1-2 minutes on my local laptop running postgres 12 and PostGIS 3.0.

In this branch I changed the code so  I let many threads working on the edges at the same time and that seems to kick off a  core dump quite often, but not every time. 

How to get it to core_dump

Run test to produce code needed (the tests will fail thats ok)
./run_test.pl --verbose --spatial_ref_sys -topology resolve_overlap_and_gap

Install the code in  an empty database

psql  -h localhost -U postgres postgres -c "create database resolve_og"
psql  -h localhost -U postgres resolve_og -c "CREATE EXTENSION dblink"
psql  -h localhost -U postgres resolve_og -c "CREATE EXTENSION postgis"
psql  -h localhost -U postgres resolve_og -c "CREATE EXTENSION postgis_topology"
psql  -h localhost -U postgres resolve_og -c "CREATE EXTENSION pg_stat_statements "

# what ever path you  have
psql  -h localhost -U postgres resolve_og -f /Users/lop/dev/github/resolve-overlap-and-gap/src/test/sql/regress/resolve_overlap_and_gap-pre.sql
psql  -h localhost -U postgres resolve_og -f /Users/lop/dev/github/postgres_execute_parallel/src/main/sql/function_execute_parallel.sql

-- Create data to test test in meter
psql  -h localhost -U postgres resolve_og -c "CREATE table test_data.overlap_gap_input_t3 AS (SELECT distinct c1, c2, c3, ST_transform(geom,25833)::Geometry(Polygon,25833) as geom from test_data.overlap_gap_input_t1)"
psql  -h localhost -U postgres resolve_og -c "CREATE index on test_data.overlap_gap_input_t3 using gist(geom)"

-- Run the test
	psql  -h localhost -U postgres resolve_og -c "CALL resolve_overlap_gap_run(
	('test_data.overlap_gap_input_t3','c1','geom',25833,true), -- TYPE resolve_overlap_data_input
	('test_topo_t3',1.0), -- TYPE resolve_overlap_data_topology
	resolve_overlap_data_clean_type_func(  -- TYPE resolve_overlap_data_clean
	49,  -- if this a polygon  is below this limit it will merge into a neighbour polygon. The area is sqare meter.
	10, -- is this is more than zero simply will called with
	null, -- _max_average_vertex_length, in meter both for utm and deegrees, this used to avoid running ST_simplifyPreserveTopology for long lines lines with few points
	2, -- IF 0 NO CHAKINS WILL BE DONE A big value here make no sense because the number of points will increaes exponential )
	10000, --edge that are longer than this value will not be touched by _chaikins_min_degrees and _chaikins_max_degrees  
	120, -- The angle has to be less this given value, This is used to avoid to touch all angles. 
	240, -- OR the angle has to be greather than this given value, This is used to avoid to touch all angles 
	40, -- The angle has to be less this given value, This is used to avoid to touch all angles. 
	320 -- OR The angle has to be greather than this given value, This is used to avoid to touch all angles 
	),
	4, -- number of threads 
	10 -- max numver of polygons in each cell 
	)"

**Comment by larsop** [comment]
In postgres server log I see this
2020-04-29 11:59:00.485 CEST [10405] LOG:  server process (PID 39391) was terminated by signal 11: Segmentation fault: 11
2020-04-29 11:59:00.485 CEST [10405] DETAIL:  Failed process was running: CALL resolve_overlap_gap_single_cell(
          'test_data.overlap_gap_input_t3','geom','c1','test_topo_t3.overlap_gap_input_t3',
          'test_topo_t3',1,25833,'true',
          '(49,10,,2,10000,120,240,2,40)',
          'test_topo_t3.overlap_gap_input_t3_job_list','test_topo_t3.overlap_gap_input_t3_grid','0103000020E96400000100000005000000825196FCD739114199A2C4F813415941825196FCD739114199A2C4F819415941584CBC056864124199A2C4F819415941584CBC056864124199A2C4F813415941825196FCD739114199A2C4F813415941',3,1);
2020-04-29 11:59:00.485 CEST [10405] LOG:  terminating any other active server processes
2020-04-29 11:59:00.485 CEST [38248] WARNING:  terminating connection because of crash of another server process
2020-04-29 11:59:00.485 CEST [38248] DETAIL:  The postmaster has commanded this server process to roll back the current transaction and exit, because another server process exited abnormally and possibly corrupted shared memory.

**Comment by strk** [comment]
On Wed, Apr 29, 2020 at 03:02:01AM -0700, Lars Aksel Opsahl wrote:
> 
> I now made branch https://github.com/larsop/resolve-overlap-and-gap/tree/postgres_core_dump
>  and a small data set that core dumped 4 off 5  times after 1-2 minutes on my local laptop running postgres 12 and PostGIS 3.0.

Can you add a Travis recipe to reproduce the crash there too ?
Create a pull request from that branch and give it those instructions.

> ./run_test.pl --verbose --spatial_ref_sys -topology resolve_overlap_and_gap

This is blocked by #12 at this moment, for me, still trying to find
a way to run those tests.

**Comment by strk** [comment]
On Wed, Apr 29, 2020 at 03:04:21AM -0700, Lars Aksel Opsahl wrote:
> In postgres server log I see this
> 2020-04-29 11:59:00.485 CEST [10405] LOG:  server process (PID 39391) was terminated by signal 11: Segmentation fault: 11
> 2020-04-29 11:59:00.485 CEST [10405] DETAIL:  Failed process was running: CALL resolve_overlap_gap_single_cell(
>           'test_data.overlap_gap_input_t3','geom','c1','test_topo_t3.overlap_gap_input_t3',
>           'test_topo_t3',1,25833,'true',
>           '(49,10,,2,10000,120,240,2,40)',
>           'test_topo_t3.overlap_gap_input_t3_job_list','test_topo_t3.overlap_gap_input_t3_grid','0103000020E96400000100000005000000825196FCD739114199A2C4F813415941825196FCD739114199A2C4F819415941584CBC056864124199A2C4F819415941584CBC056864124199A2C4F813415941825196FCD739114199A2C4F813415941',3,1);

Further simplifying the testcase is the thing to do here.
Having it tested on Travis would also help checking against
different versions of PostGIS and PostgreSQL.

**Comment by larsop** [comment]
Now Travis core dumps and my local lap top to when I do run_test.pl --verbose --spatial_ref_sys -topology resolve_overlap_and_gap' 

https://travis-ci.org/github/larsop/resolve-overlap-and-gap/builds/686869828

**Comment by larsop** [comment]
Here is some more info about core dumps that I am only able to trigger with bigger datasets . Here in [in step 2](https://github.com/larsop/resolve-overlap-and-gap/commit/1a7cedfe93f5f55feea6614745b9d98ce5cca30f) where I start to merge single cell into the main result Topology layer when run at very high CPU load . At this step I am only adding lines that are not crossing any cell borders. I have tested that checed that none of faces are touching the cell borders at this stage. (tested a tie it did not core dump)

```
SELECT f.* FROM 
   topo_sr16_mdata_05.face f,
   SELECT ST_ExteriorRing(geo) as  geo from topo_sr16_mdata_05.trl_2019_test_segmenter_mindredata_grid) g
   where ST_Intersects(f.mbr,g.geo)
   ;
    face_id | mbr 
   -------+-----
   (0 rows)

```
The server I test on has 32 cores. For instance if run in 28 parallel threads it core dump when starting at 'step 2'  in 90 % percent of the times. If I reduce it to 18 parallel threads it fails maybe around 10% of the times or maybe less.

**Comment by strk** [comment]
I do can reproduce the crash with:
> resolve-overlap-and-gap(postgres_core_dump)] make check

Looking a it !

**Comment by strk** [comment]
I've tried changing the test file to run with a _single_ thread, with this patch:
```
diff --git a/src/test/sql/regress/resolve_overlap_and_gap.sql b/src/test/sql/regress/resolve_overlap_and_gap.sql
index edd5cf4..8e8612a 100644
--- a/src/test/sql/regress/resolve_overlap_and_gap.sql
+++ b/src/test/sql/regress/resolve_overlap_and_gap.sql
@@ -57,7 +57,9 @@ resolve_overlap_data_clean_type_func(  -- TYPE resolve_overlap_data_clean
 40, -- The angle has to be less this given value, This is used to avoid to touch all angles. 
 320 -- OR The angle has to be greather than this given value, This is used to avoid to touch all angles 
 ),
-5,4);
+1, -- number of threads
+4  -- max number of polygons in each cell
+);
 
 SELECT 'utm_check_failed_lines', count(geo) from test_topo_t3.overlap_gap_input_t3_no_cut_line_failed;
 
@@ -89,6 +91,6 @@ null, -- _max_average_vertex_length, in meter both for utm and deegrees, this us
 40, -- The angle has to be less this given value, This is used to avoid to touch all angles.
 320 -- OR The angle has to be greather than this given value, This is used to avoid to touch all angles
 ),
-4, -- number of threads
+1, -- number of threads
 10 -- max numver of polygons in each cell
-);
\ No newline at end of file
+);
```

The `make check` call, with test script so modified, takes a very very long time and high CPU load, as  you observed, but what concerns me is that the PostgreSQL logs do contain ERROR lines, which I'd expect to STOP the operations, which instead keep going on:

```
2020-05-15 16:38:56.762 CEST [30631] postgres@nibio_reg ERROR:  update or delete on table "edge_data" violates foreign key constraint "next_left_ed
ge_exists" on table "edge_data"
2020-05-15 16:38:56.762 CEST [30631] postgres@nibio_reg DETAIL:  Key (edge_id)=(67) is still referenced from table "edge_data".
2020-05-15 16:38:56.762 CEST [30631] postgres@nibio_reg STATEMENT:  CALL resolve_overlap_gap_single_cell(
          'test_data.overlap_gap_input_t2','geom','c1','test_topo_t2.overlap_gap_input_t2',
          'test_topo_t2',1e-05,4258,'false',
          '(49,0,,0,10000,120,240,0,40)',
          'test_topo_t2.overlap_gap_input_t2_job_list','test_topo_t2.overlap_gap_input_t2_grid','0103000020A21000000100000005000000970992439B331F40
340CF422950E4E40970992439B331F40AF14B064BA364E4041A4B22DDA331F40AF14B064BA364E4041A4B22DDA331F40340CF422950E4E40970992439B331F40340CF422950E4E40',4
,26967);
```

See, 2 minutes later, another such error:
```
2020-05-15 16:40:22.177 CEST [3960] postgres@nibio_reg ERROR:  update or delete on table "edge_data" violates foreign key constraint "next_left_edge_exists" on table "edge_data"
2020-05-15 16:40:22.177 CEST [3960] postgres@nibio_reg DETAIL:  Key (edge_id)=(67) is still referenced from table "edge_data".
2020-05-15 16:40:22.177 CEST [3960] postgres@nibio_reg STATEMENT:  CALL resolve_overlap_gap_single_cell(
          'test_data.overlap_gap_input_t2','geom','c1','test_topo_t2.overlap_gap_input_t2',
          'test_topo_t2',1e-05,4258,'false',
          '(49,0,,0,10000,120,240,0,40)',
          'test_topo_t2.overlap_gap_input_t2_job_list','test_topo_t2.overlap_gap_input_t2_grid','0103000020A21000000100000005000000970992439B331F40340CF422950E4E40970992439B331F40AF14B064BA364E4041A4B22DDA331F40AF14B064BA364E4041A4B22DDA331F40340CF422950E4E40970992439B331F40340CF422950E4E40',4,28762);
```

Shouldn't operations STOP on first error ? Or how do you otherwise handle those errors ?

**Comment by strk** [comment]
Just a side note: I still get `failed, '.' is no longer in @INC` in the `postgres_core_dump` branch, how about rebasing it to current master, where that problem was fixed ?

**Comment by strk** [comment]
Sorry, I was wrong, it's NOT fixed in master branch (it's fixed when using `make check`, but the direct call needs to be fully qualified). So what you can use to run warning is:

> ./run_test.pl --verbose --extension -topology $PWD/resolve_overlap_and_gap

Can you reproduce the segfault with single thread ? It takes a long time so it would still be good to reduce the dataset, if possible. Of course as long as we use threads it'll be harder to tell if data reduction can still reproduce the problem.

**Comment by strk** [comment]
I'm attaching a zip file containing 2 files: crash_dump.sql and crash.sql
The first file is a dump of the starting database condition, the second is the script that allows me to reproduce the crash, 100%.

The plan is to further reduce the dump to really only contain what's needed.
This means getting rid of unused tables and schema for a start, but later also removing unneeded records from the input tables.

Such package will be a good one to post on upstream PostGIS trac for further analisys, if we find out it's really PostGIS triggering the crash (the only other possibility I can think of would be dblink)

[crash.zip](https://github.com/larsop/resolve-overlap-and-gap/files/4635743/crash.zip)

**Comment by larsop** [comment]
Thats very good. 

I may be able to rewrite this code using GNU parallel to check if the same happens, but I have been using dblink with other projects like https://github.com/larsop/find-overlap-and-gap and have not seen this problem here so far.

**Comment by larsop** [comment]
Last night a another Postgres server core dumped it had been running a job for more than 3 days with 50 parallel threads **not involving Postgis** but just Postgres code where we use a lot of temp tables and Postgres core dumped here also.

```
>WARNING:  terminating connection because of crash of another server process
< postgres, resolve_og, 2020-05-21 20:59:07.525 CEST >DETAIL:  The postmaster has commanded this server process to roll back the current transaction and exit, because another server process exited abnormally and possibly corrupted shared memory.
```

**Comment by larsop** [comment]
Could not find any "Out of Memory" in any logs. So it seems like it is not related to over commit https://www.postgresql.org/docs/current/kernel-resources.html#LINUX-MEMORY-OVERCOMMIT

**Comment by strk** [comment]
Running crash.sql script with a single thread (1 as the "number of threads" argument) results in NO crash, so this seems related to concurrency as well. It takes ~7m30s to complete in that way. I'll try to further reduce the test so to take less time, and then obtain a stacktrace with valgrind

**Comment by strk** [comment]
I've reduced the testcase so now it crashes in ~6 seconds with 10 threads and in ~14 seconds with 2 threads. Still does NOT crash with a single thread (which completes in ~40 seconds).
[crash.zip](https://github.com/larsop/resolve-overlap-and-gap/files/4803215/crash.zip)

Valgrinding the run was not straightforward as I was hoping for, due to the dblink usage (can't use single user mode, when we're actually trying to connect again!)

**Comment by larsop** [comment]
Thats good. 

Do think resolving this bug can resolve https://github.com/larsop/resolve-overlap-and-gap/issues/16, when using blocking grid cells ?

Because I have I tried block very big areas and also tested based making block based on input polygons  and is neighbour and that did not help either.

**Comment by strk** [comment]
Memory errors can cause any kind of issue but I don't think #16 is related to this.
Attaching a correct version of a further reduction of the dump (previous one was a duplicated of previous)
[crash2.zip](https://github.com/larsop/resolve-overlap-and-gap/files/4803569/crash2.zip)

**Comment by strk** [comment]
Backtrace:
```
Core was generated by `postgres: 12/main: postgres nibio [local] CALL                                '.
Program terminated with signal SIGSEGV, Segmentation fault.
#0  0x00007f14389b5a16 in _box2d_to_lwgeom (bbox=0x0, srid=25833) at postgis_topology.c:129
129       POINTARRAY *pa = ptarray_construct(0, 0, 2);
(gdb) bt
#0  0x00007f14389b5a16 in _box2d_to_lwgeom (bbox=0x0, srid=25833) at postgis_topology.c:129
#1  0x00007f14389b5aad in _box2d_to_hexwkb (bbox=<optimized out>, srid=<optimized out>) at postgis_topology.c:148
#2  0x00007f14389b76ee in cb_updateFacesById (topo=0x55939bd38c50, faces=0x7ffeccba6fb0, numfaces=2) at postgis_topology.c:1924
#3  0x00007f14389fbb7f in lwt_ChangeEdgeGeom (topo=topo@entry=0x55939bd38cc0, edge_id=<optimized out>, edge_id@entry=188, 
    geom=geom@entry=0x55939beba660) at lwgeom_topo.c:3515
#4  0x00007f14389fc3fe in _lwt_AddPoint (topo=0x55939bd38cc0, point=<optimized out>, tol=<optimized out>, findFace=<optimized out>, 
    moved=0x7ffeccba7154) at lwgeom_topo.c:5160
#5  0x00007f14389fcccc in _lwt_AddLineEdge (handleFaceSplit=1, tol=1, edge=0x55939bef3668, topo=0x55939bd38cc0) at lwgeom_topo.c:5336
#6  _lwt_AddLine (topo=topo@entry=0x55939bd38cc0, line=line@entry=0x55939bd232c0, tol=tol@entry=1, nedges=nedges@entry=0x7ffeccba727c, 
    handleFaceSplit=handleFaceSplit@entry=1) at lwgeom_topo.c:5769
#7  0x00007f14389fe11e in lwt_AddLine (topo=topo@entry=0x55939bd38cc0, line=line@entry=0x55939bd232c0, tol=tol@entry=1, 
    nedges=nedges@entry=0x7ffeccba727c) at lwgeom_topo.c:5800
#8  0x00007f14389bdbff in TopoGeo_AddLinestring (fcinfo=0x55939bd56380) at postgis_topology.c:4775
```

**Comment by larsop** [comment]
Good, Is it resulting in a empty bb and that is causing the crash ?

**Comment by strk** [comment]
judging by the backtrace the segfault is due to an invalid topology, so issue #16 might indeed be the _cause_ for this, not a symptom. PostGIS should still NOT crash in this case. I filed https://trac.osgeo.org/postgis/ticket/4706#ticket to fix it upstream

**Comment by strk** [comment]
Ok, the upstream bug is fixed: https://trac.osgeo.org/postgis/ticket/4706

Now, on _this_ side of things, knowing the cause is a corrupted topology, the code handling the exception thrown by backend should now stop trying rather than assuming the exception is due to some robustness issue and changing the tolerance (which is what happens now).

This retrying, I think, may be one cause of issue #16 (desperately trying with different tolerance values when the actual cause is a corruption of the topology, which should instead trigger a SHOW STOP)

**Comment by larsop** [comment]
Very nice that you fixed this. 

Trigger a "SHOW STOP", means that I should do a roll back ?

**Comment by strk** [comment]
I'm more inclined to COMMIT and STOP to see what kind of invalidity was produced, rather than ROLLBACK.

When an invalidity is introduced, it means we've either failed to protect concurrent update of rows OR there's a bug in PostGIS Topology, so we want to stop as soon as possible and try to re-create that scenario to see what have happened.

**Comment by larsop** [comment]
Seems like Posgres still core dumps even after running with latest compiled code from GitHub

 POSTGIS="3.1.0dev 3.1.0alpha1-137-ga2557be" [EXTENSION] PGSQL="120" GEOS="3.8.1-CAPI-1.13.3" PROJ="7.0.1" LIBXML="2.9.1" LIBJSON="0.11" TOPOLOGY

> NOTICE:  cell 1632 cell_job_type 2, has_edges t, _loop_number 1
> NOTICE:  select id from test_topo_ar5.ar5_2019_komm_flate_job_list where cell_geo = '0103000020A21000000100000005000000A8AB89D8E0BD27404216F00D74F74F40A8AB89D8E0BD2740F64497238A025040E
> 9787D17932C2840F64497238A025040E9787D17932C28404216F00D74F74F40A8AB89D8E0BD27404216F00D74F74F40' 
> NOTICE:  enter at timeofday:Mon Jun 22 20:08:30.563749 2020 CEST for layer test_topo_ar5_, with _cell_job_type 2 and box id 3194 .
> NOTICE:  start work at timeofday:Mon Jun 22 20:08:30.566736 2020 CEST for layer test_topo_ar5_3194, _topology_snap_tolerance 1e-06, with _cell_job_type 2 and min_length_line 0.049s, (_
> clean_info).chaikins_max_degrees) 240
> NOTICE:  start work at timeofday:Mon Jun 22 20:08:30.553633 2020 CEST for layer test_topo_ar5_2787, _topology_snap_tolerance 1e-06, with _cell_job_type 2 and min_length_line 0.049s, (_
> clean_info).chaikins_max_degrees) 240
> NOTICE:  cell 2787 cell_job_type 2, has_edges t, _loop_number 1
> NOTICE:  select id from test_topo_ar5.ar5_2019_komm_flate_job_list where cell_geo = '0103000020A21000000100000005000000E10EDF1F01B72440F64497238A025040E10EDF1F01B72440CB7E36405A0950402
> 2DCD25EB3252540CB7E36405A09504022DCD25EB3252540F64497238A025040E10EDF1F01B72440F64497238A025040' 
> NOTICE:  select id from test_topo_ar5.ar5_2019_komm_flate_job_list where cell_geo = '0103000020A21000000100000005000000E10EDF1F01B72440642DE3ACCC354E40E10EDF1F01B724400EA121E66C434E402
> 2DCD25EB32525400EA121E66C434E4022DCD25EB3252540642DE3ACCC354E40E10EDF1F01B72440642DE3ACCC354E40' 
> < , 2020-06-22 20:08:30 CEST, , 2020-06-22 20:08:30.569 CEST >LOG:  server process (PID 36766) was terminated by signal 11: Segmentation fault
> < , 2020-06-22 20:08:30 CEST, , 2020-06-22 20:08:30.569 CEST >DETAIL:  Failed process was running: CALL resolve_overlap_gap_single_cell(
>           'org_ar5arsversjon.ar5_2019_komm_flate','geo','sl_sdeid','test_topo_ar5.ar5_2019_komm_flate',
>           'test_topo_ar5',1e-06,4258,'false',
>           '(49,0,500,0,140,120,240,0,35)',
>           'test_topo_ar5.ar5_2019_komm_flate_job_list','test_topo_ar5.ar5_2019_komm_flate_grid','0103000020A210000001000000050000006F483491C0C42A40BEEA23EFDCAC50406F483491C0C42A409324C30BADB35040B01528D072332B409324C30BADB35040B01528D072332B40BEEA23EFDCAC50406F483491C0C42A40BEEA23EFDCAC5040',2,1);
> < , 2020-06-22 20:08:30 CEST, , 2020-06-22 20:08:30.569 CEST >LOG:  terminating any other active server processes
> < postgres, 2020-06-22 20:08:30 CEST, resolve_cha, 2020-06-22 20:08:30.570 CEST >WARNING:  terminating connection because of crash of another server process at character 296
> < postgres, 2020-06-22 20:08:30 CEST, resolve_cha, 2020-06-22 20:08:30.570 CEST >DETAIL:  The postmaster has commanded this server process to roll back the current transaction and exit, because another server process exited abnormally and possibly corrupted shared memory.
> < postgres, 2020-06-22 20:08:30 CEST, resolve_cha, 2020-06-22 20:08:30.570 CEST >HINT:  In a moment you should be able to reconnect to the database and repeat your command.
> < postgres, 2020-06-22 20:08:30 CEST, resolve_cha, 2020-06-22 20:08:30.570 CEST >WARNING:  terminating connection because of crash of another server process at character 296
> < postgres, 2020-06-22 20:08:30 CEST, resolve_cha, 2020-06-22 20:08:30.570 CEST >DETAIL:  The postmaster has commanded this server process to roll back the current transaction and exit, because another server process exited abnormally and possibly corrupted shared memory.
> < postgres, 2020-06-22 20:08:30 CEST, resolve_cha, 2020-06-22 20:08:30.570 CEST >HINT:  In a moment you should be able to reconnect to the database and repeat your command.
> < postgres, 2020-06-22 20:08:30 CEST, resolve_cha, 2020-06-22 20:08:30.570 CEST >WARNING:  terminating connection because of crash of another server process at character 296
> < postgres, 2020-06-22 20:08:30 CEST, resolve_cha, 2020-06-22 20:08:30.570 CEST >DETAIL:  The postmaster has commanded this server process to roll back the current transaction and exit, because another server process exited abnormally and possibly corrupted shared memory.
> < postgres, 2020-06-22 20:08:30 CEST, resolve_cha, 2020-06-22 20:08:30.570 CEST >HINT:  In a moment you should be able to reconnect to the database and repeat your command.
> < postgres, 2020-06-22 20:08:30 CEST, resolve_cha, 2020-06-22 20:08:30.570 CEST >WARNING:  terminating connection because of crash of another server process
> < postgres, 2020-06-22 20:08:30 CEST, resolve_cha, 2020-06-22 20:08:30.570 CEST >DETAIL:  The postmaster has commanded this server process to roll back the current transaction and exit, because another server process exited abnormally and possibly corrupted shared memory.
> 
> 
> 
>

**Comment by larsop** [comment]
It's much more stable than before and I now did test with 28 parallel threads without core dump one a medium sized dataset. (ar50).

**Comment by strk** [comment]
Can you provide a small test dataset like the one I attached in https://github.com/larsop/resolve-overlap-and-gap/issues/1#issuecomment-646527313 ? I'll try to reproduce but like before this may be due to robustness problems in presence of invalid topologies so issue #18 becomes more and more important to deal with these issues

**Comment by strk** [comment]
It looks like the first attachment of mine is enough to reproduce (crash.zip).
Here's a new backtrace:
```
Core was generated by `postgres: 12/main: postgres nibio [local] CALL                                '.
Program terminated with signal SIGSEGV, Segmentation fault.
#0  lwline_free (line=0x18000000040000) at lwline.c:71
71              if ( line->bbox )
(gdb) bt
#0  lwline_free (line=0x18000000040000) at lwline.c:71
#1  0x00007f83132b3f1d in _lwt_release_edges (edges=0x55fd20092a18, num_edges=<optimized out>) at lwgeom_topo.c:455
#2  0x00007f83132b6d49 in _lwt_MakeRingShell (num_signed_edge_ids=<optimized out>, signed_edge_ids=0x55fd200a0c38, topo=0x55fd1fe70ae0)
    at lwgeom_topo.c:1774
#3  _lwt_AddFaceSplit (topo=topo@entry=0x55fd1fe70ae0, sedge=<optimized out>, face=<optimized out>, mbr_only=mbr_only@entry=1) at lwgeom_topo.c:1892
#4  0x00007f83132b79e6 in _lwt_AddEdge (topo=topo@entry=0x55fd1fe70ae0, start_node=<optimized out>, end_node=8341, geom=geom@entry=0x55fd20010dc8, 
    skipChecks=skipChecks@entry=0, modFace=modFace@entry=1) at lwgeom_topo.c:2684
#5  0x00007f83132bcf82 in _lwt_AddLineEdge (handleFaceSplit=1, tol=1, edge=0x55fd20010dc8, topo=0x55fd1fe70ae0) at lwgeom_topo.c:5495
#6  _lwt_AddLine (topo=topo@entry=0x55fd1fe70ae0, line=line@entry=0x55fd1fe5be80, tol=tol@entry=1, nedges=nedges@entry=0x7fffb4b55bfc, 
    handleFaceSplit=handleFaceSplit@entry=1) at lwgeom_topo.c:5781
#7  0x00007f83132be16e in lwt_AddLine (topo=topo@entry=0x55fd1fe70ae0, line=line@entry=0x55fd1fe5be80, tol=tol@entry=1, 
    nedges=nedges@entry=0x7fffb4b55bfc) at lwgeom_topo.c:5812
#8  0x00007f831327dbff in TopoGeo_AddLinestring (fcinfo=0x55fd1fe606b0) at postgis_topology.c:4775
```

**Comment by larsop** [comment]
> It looks like the first attachment of mine is enough to reproduce (crash.zip).
> Here's a new backtrace:
> 

Great, the I test I did was running for hours on big server glad I don't have to start to reduce that test case.

**Comment by strk** [comment]
The other very useful thing to do is fixing #17 because otherwise `client_min_messages` is useless (interesting messages will never get shown by the client...)

**Comment by strk** [comment]
Debug logs:
```
2020-06-26 09:55:26.514 CEST [1078] LOG:  server process (PID 19430) was terminated by signal 11: Segmentation fault
2020-06-26 09:55:26.514 CEST [1078] DETAIL:  Failed process was running: CALL resolve_overlap_gap_single_cell(
          'test_data.overlap_gap_input_t4','geom','c1','test_topo_t4.overlap_gap_input_t4',
          'test_topo_t4',1,25833,'true',
          '(49,10,,2,10000,120,240,2,40)',
          'test_topo_t4.overlap_gap_input_t4_job_list','test_topo_t4.overlap_gap_input_t4_grid','0103000020E96400000100000005000000AFB794D4EFCA0D41E2613E0E77E35841AFB794D4EFCA0D41E2613E0E7DE35841855196FC373A1141E2613E0E7DE35841855196FC373A1141E2613E0E77E35841AFB794D4EFCA0D41E2613E0E77E35841',3,1);
2020-06-26 09:55:26.514 CEST [1078] LOG:  terminating any other active server processes
```

Now, the crashing backend is PID 19430, running as last statement:
```
2020-06-26 09:55:26.277 CEST [19430] postgres@nibio CONTEXT:  SQL statement "select topology.TopoGeo_addLinestring('test_topo_t4','0102000020E9640000020
000003ED86BD30AFF0D41E2613E0E75E35841F00F9ECD3EFF0D41E2613E8E7CE35841',1)"
```

The above TopoGeo_addLinestring statement resulted in:
```
2020-06-26 09:55:26.277 CEST [19430] postgres@nibio DEBUG:  [lwgeom_topo.c:_lwt_AddFaceSplit:1876] getRingEdges returned 6259 edges
2020-06-26 09:55:26.277 CEST [19430] postgres@nibio DEBUG:  [lwgeom_topo.c:_lwt_AddFaceSplit:1888] Edge -14861 split face 0 (mbr_only:1)
```

So the new edge added (`0103000020E96400000100000005000000AFB794D4EFCA0D41E2613E0E77E35841AFB794D4EFCA0D41E2613E0E7DE35841855196FC373A1141E2613E0E7DE35841855196FC373A1141E2613E0E77E35841AFB794D4EFCA0D41E2613E0E77E35841`) was found to close a new ring composed by 6259 edges, and was found to be contained in the universe face, thus "splitting" it.. The function `_lwt_MakeRingShell` is invoked with the identifiers of these 6259 edges to construct a polygon from them. It queries the backend to obtain the geometries of these edges, but it receives a different number of edges than the expected one, when querying for those 6259 edges, for some reason. Then it crashes while trying to release the memory associated to the returned set of edges (which seems to be a valid pointer). I'll try reducing this case as well.

**Comment by strk** [comment]
The problem is not always easy to reproduce, as enabling debug statements makes the problem go away. Most likely due to the timing in which the 2 threads I'm using happen to do their things. THis is what I described in https://github.com/larsop/resolve-overlap-and-gap/issues/16#issuecomment-650058290 ... Multi-process is really hard to deal with.

**Comment by strk** [comment]
I think I found the culprit, although I'm still not sure how to reproduce.
The function extracting an edge's ring returns a list of edge identifiers. These edge identifiers are then used to query the database for the edges geometries (removing duplicates). The query returns less edges then expected:

> NOTICE:  failed after trying with tolerance 6.75 : XX000 message: Unexpected error: 6461 edges found when expecting 6462 detail :  hint   :  context: SQ
L statement "SELECT ARRAY(SELECT topology.TopoGeo_addLinestring('test_topo_t4','0102000020E9640000040000003B28C645D88B1A41366677CC3F675B413B28C645B08B1A
4123A36AD93E675B413B28C645608B1A41FD1C51F33C675B413B28C645388B1A41EA5944003C675B41',6.75))"

The PostGIS Topology code performing cleanup operations upon catching that error erroneously attempts to release memory associated with the expected number of edges, thus cleaning too much (triggering the segfault).

I've filed this upstream to https://trac.osgeo.org/postgis/ticket/4709

**Comment by strk** [comment]
@larsop I've pushed a fix for the issue, although I wasn't yet able to create a proper testcase for it. There must be a condition in which GetRingEdges returns more edges than needed to actually create a ring, and include an unneeded edge (I guess it must be a dangling final edge). While working on these scenarios I found more ones, like returning all existing edges but them not forming a link... lots of issues coming up when topology is invalid. I'd focus on my side on improving the robustness and user messages so that your code has a way to catch as early as possible these messages and STOP execution (#18)

**Comment by larsop** [comment]
I got the latest changes from master and compiled it and installed(make clean, make, make install) with out any errors, but I was not able to verify that I used  the latest when checking the extensions.

```
ALTER EXTENSION postgis UPDATE TO "3.1.0dev";
NOTICE:  00000: version "3.1.0dev" of extension "postgis" is already installed
LOCATION:  ExecAlterExtensionStmt, extension.c:2991
ALTER EXTENSION

ALTER EXTENSION postgis_topology UPDATE TO "3.1.0dev";
NOTICE:  00000: version "3.1.0dev" of extension "postgis_topology" is already installed
LOCATION:  ExecAlterExtensionStmt, extension.c:2991
ALTER EXTENSION

```

Anyway I did run a test and it failed at stage 2, where we insert rows into final topology layer, but in stage 2 each thread are only working with lines inside each cell so no lines inserted are connected with line handled by another thread.

```
< postgres, 2020-06-26 18:54:09 CEST, resolve_cha, 2020-06-26 18:54:09.930 CEST >WARNING:  terminating connection because of crash of another server process
< postgres, 2020-06-26 18:54:09 CEST, resolve_cha, 2020-06-26 18:54:09.930 CEST >DETAIL:  The postmaster has commanded this server process to roll back the current transaction and exit, 
because another server process exited abnormally and possibly corrupted shared memory.
< postgres, 2020-06-26 18:54:09 CEST, resolve_cha, 2020-06-26 18:54:09.930 CEST >HINT:  In a moment you should be able to reconnect to the database and repeat your command.
< postgres, 2020-06-26 18:54:09 CEST, resolve_cha, 2020-06-26 18:54:09.930 CEST >CONTEXT:  SQL statement "WITH faces AS ( SELECT face_id FROM "test_topo_ar50_t7".face WHERE mbr && $1 ORD
ER BY ST_Area(mbr) ASC ) SELECT face_id FROM faces WHERE _ST_Contains(topology.ST_GetFaceGeometry('test_topo_ar50_t7', face_id), $1) LIMIT 1"
        SQL statement "SELECT ARRAY(SELECT topology.TopoGeo_addLinestring('test_topo_ar50_t7',r.geom,1)) FROM 
                                         (SELECT geom from test_topo_ar50_t7.edge_data_tmp_249 order by is_closed desc, num_points desc) as r"
        PL/pgSQL function resolve_overlap_gap_single_cell(character varying,character varying,character varying,character varying,character varying,double precision,integer,boolean,resol
ve_overlap_data_clean_type,character varying,character varying,geometry,integer,integer) line 372 at EXECUTE
< , 2020-06-26 18:54:09 CEST, , 2020-06-26 18:54:09.931 CEST >WARNING:  terminating connection because of crash of another server process
< , 2020-06-26 18:54:09 CEST, , 2020-06-26 18:54:09.931 CEST >DETAIL:  The postmaster has commanded this server process to roll back the current transaction and exit, because another ser
ver process exited abnormally and possibly corrupted shared memory.
< , 2020-06-26 18:54:09 CEST, , 2020-06-26 18:54:09.931 CEST >HINT:  In a moment you should be able to reconnect to the database and repeat your command.
< , 2020-06-26 18:54:10 CEST, , 2020-06-26 18:54:10.029 CEST >LOG:  all server processes terminated; reinitializing
< , 2020-06-26 18:54:16 CEST, , 2020-06-26 18:54:16.166 CEST >LOG:  database system was interrupted; last known up at 2020-06-26 18:47:51 CEST
< , 2020-06-26 18:54:16 CEST, , 2020-06-26 18:54:16.516 CEST >LOG:  database system was not properly shut down; automatic recovery in progress
< , 2020-06-26 18:54:35 CEST, , 2020-06-26 18:54:35.943 CEST >LOG:  redo starts at 20/45DE7070
< , 2020-06-26 18:56:08 CEST, , 2020-06-26 18:56:08.518 CEST >LOG:  redo done at 20/48B740C8
< , 2020-06-26 18:56:09 CEST, , 2020-06-26 18:56:09.540 CEST >LOG:  database system is ready to accept connections

```

(this test was done before I added code show error from postgres execute parallel and stop if topology validations indicates errors)

**Comment by strk** [comment]
On Sat, Jun 27, 2020 at 11:54:41PM -0700, Lars Aksel Opsahl wrote:

> ALTER EXTENSION postgis UPDATE TO "3.1.0dev";
> NOTICE:  00000: version "3.1.0dev" of extension "postgis" is already installed
> LOCATION:  ExecAlterExtensionStmt, extension.c:2991
> ALTER EXTENSION
> 
> ALTER EXTENSION postgis_topology UPDATE TO "3.1.0dev";
> NOTICE:  00000: version "3.1.0dev" of extension "postgis_topology" is already installed
> LOCATION:  ExecAlterExtensionStmt, extension.c:2991
> ALTER EXTENSION

Check with `SELECT postgis_full_version()` if you are upgraded or not.
Check out `postgis_extensions_upgrade()` too, may be handy for these
cases.

> Anyway I did run a test and it failed at stage 2, where we insert rows into final topology layer, but in stage 2 each thread are only working with lines inside each cell so no lines inserted are connected with line handled by another thread.
> 
> ```
> < postgres, 2020-06-26 18:54:09 CEST, resolve_cha, 2020-06-26 18:54:09.930 CEST >WARNING:  terminating connection because of crash of another server process

To better debug crashes, set the following line:

    pg_ctl_options = '-c'

In the /etc/postgresql/12/main/pg_ctl.conf file
(or wherever is your pg_ctl.conf file for your cluster)

Then issue:

    pg_ctlcluster 12 main reload

This should leave a "core" file in the DATADIR, for example:

    /var/lib/postgresql/12/main/core

A backtrace from the core can be obtained with:

    gdb /usr/lib/postgresql/12/bin/postgres core

Then from the prompt (gdb):

    bt
  
But... too many crashes, please let's focus on one at the time :)

Right now I'm chasing the one triggered by crash2.zip, which
is still lacking a testcase to secure the fix upstream (I'll be
working on that).

It would help if you could verify the crash is gone with the dump
(crash_dump2.sql) and call (crash.sql) included there.

    createdb test
    psql -f crash_dump2.sql test
    psql -f crash.sql test

Rather than a crash, you should get errors due to topology being
invalid/corrupted. You can try multiple times, changing the
number of threads in the crash.sql file.

Based on this we should aim at UPGRADING
the resolve-overlap-and-gap code in the database (after restoring
the crash_dump.sql script) to see the effects of that.

It would help further simplifying the tester code having the
resolve-overlap-and-gap code easily installable/upgradeable
(so we can even remove it from the dump). See #19

**Comment by larsop** [comment]
> Check with `SELECT postgis_full_version()` if you are upgraded or not. Check out `postgis_extensions_upgrade()` too, may be handy for these cases.

I did run the postgis_extensions_upgrade() 

POSTGIS="3.1.0dev 3.1.0alpha1-137-ga2557be" [EXTENSION] PGSQL="120" GEOS="3.8.1-CAPI-1.13.3" PROJ="7.0.1" LIBXML="2.9.1" LIBJSON="0.11" TOPOLOGY
(1 row)

And then did test where with your sql and went ok first time, but when upgrading to 10 parallel threads it core dumped.

DETAIL:  The postmaster has commanded this server process to roll back the current transaction and exit, because another server process exited abnormally and possibly corrupted shared memory.
HINT:  In a moment you should be able to reconnect to the database and repeat your command.

I will add the extra debug flag retry the test.

**Comment by strk** [comment]
On Mon, Jun 29, 2020 at 01:15:17AM -0700, Lars Aksel Opsahl wrote:
> > Check with `SELECT postgis_full_version()` if you are upgraded or not. Check out `postgis_extensions_upgrade()` too, may be handy for these cases.
> 
> I did run the postgis_extensions_upgrade() 
> 
> POSTGIS="3.1.0dev 3.1.0alpha1-137-ga2557be" [EXTENSION] PGSQL="120" GEOS="3.8.1-CAPI-1.13.3" PROJ="7.0.1" LIBXML="2.9.1" LIBJSON="0.11" TOPOLOGY
> (1 row)

This means you are at a2557be05fba0ed47ddd3dbb7dcf83b334581e65 which
is the fix for ST_ChangeEdgeGeom corruption handling.

In a later commit (96615d90e6d5e4716310eca25647ee373a8535f8) there's
another crasher fix. Latest PostGIS is f9db63e9dc082534798861cdfed5c1429c36406f

**Comment by larsop** [comment]
With version 

POSTGIS="3.1.0dev 3.1.0alpha1-144-g50796cc" [EXTENSION] PGSQL="120" GEOS="3.8.1-CAPI-1.13.3" PROJ="7.0.1" LIBXML="2.9.1" LIBJSON="0.11" TOPOLOGY

it seems much more stable, with the old version it crashed at my first try with 10 threads and but with the new one I have now run 5 times with no core dump yet.

**Comment by strk** [comment]
Can you find interesting errors in the logs ? I'd expect those errors to still raise (just not a crash, but an errors suggesting topology is corrupted)

**Comment by larsop** [comment]
This crash code was made without any blocking off threads to provoke  a crash and that will trigger invalid topology, so your fixes for crashing seems work good. Very nice work.  So I close this issue now and if we get new crash we may reopen it again.
